### DISCUSSION TITLE
Building an AST

### DISCUSSION THREAD

**Discussion by The-Futurist** [antlr/antlr4#4590] [Q&A]
I'm trying fathom the roles of Walker, Visitor and Listener, not yet 100% clear on the intended use of these. My opening question though is, I'm developing a compiler (I did [the grammar](https://github.com/Steadsoft/imperium/blob/main/grammar/Antlr/Imperium.g4) a year ago and I'm returning to the project) and I am inclined to (somehow) traverse the parse tree generated by the generated parser code and create a distinct, new tree - AST.

My grammar is rooted in a rule named `translationUnit` so the generated parser has this

```cs
	[RuleVersion(0)]
	public TranslationUnitContext translationUnit() {
		TranslationUnitContext _localctx = new TranslationUnitContext(Context, State);
		EnterRule(_localctx, 0, RULE_translationUnit);
```

My little test app has this:

```cs
            NestedSourceReader reader = new NestedSourceReader("grammar_test.ipl", regex, ReadFile);

            AntlrInputStream inputStream = new AntlrInputStream(reader);
            ImperiumLexer imperium_lexer = new ImperiumLexer(inputStream);

            imperium_lexer.SetLexicon("en");

            CommonTokenStream commonTokenStream = new CommonTokenStream(imperium_lexer);
            CustomParser imperium_parser = new CustomParser(commonTokenStream);

            imperium_parser.BuildParseTree = true;

            var tree = imperium_parser.translationUnit();
```
(`NestedSourceReader` is just a small helper class that implements `TextReader`). 

So I get my parse tree in the `tree` variable.

In my mind I think I want to now take _that_ generated tree and traverse it and in so doing so gradually build a new tree, one that represents my AST. Then with that AST I can consume that and begin the compiler proper.

To do this I started to create a class:

![image](https://github.com/antlr/antlr4/assets/12262952/f80ceaa9-d65b-45fc-8f4e-a5a6fafd5c16)

Which would seem to end up with a large number of methods. These methods each create a new `AstNode` which is the base class for all of the specific nodes I'd be creating in my new AST.

So my question is, is this reasonable? am I doing this in a sensible way or have I misunderstood something basic, is there a better approach to all this?

**Comment by kaby76** [comment]
> [I]s this reasonable?

Yes. That is a correct application of the [Visitor](https://github.com/antlr/antlr4/blob/380ce4b8b1658df16ada45e1d56d5aa476052376/runtime/CSharp/src/Tree/IParseTreeVisitor.cs) interface. A visitor returns values, which in your case is an AST. An example of this conversion is for [XPath expressions in Trash](https://github.com/kaby76/Domemtech.Trash/blob/79a1f1bd2a1b1255b412105aaf0a1124dec6e0ff/src/ParseTreeEditing/org/eclipse/wst/xml/xpath2/processor/internal/InternalXPathParser.cs#L64). This conversion is done because the XPath engine defines an intermediate language (a tree) that is different from the parse tree. You will need to return an AST value for every parse tree node type in the visitor.

A [Listener](https://github.com/antlr/antlr4/blob/380ce4b8b1658df16ada45e1d56d5aa476052376/runtime/CSharp/src/Tree/IParseTreeListener.cs) is a tree walker that by default a DFS tree traversal where EnterXXXXX() and ExitXXXXX() methods are called for node type XXXXX as the tree is traversed. Unlike a Visitor, which has "built-in" walker, a Listener requires a [Walker to walk the tree](https://github.com/antlr/antlr4/blob/380ce4b8b1658df16ada45e1d56d5aa476052376/runtime/CSharp/src/Tree/ParseTreeWalker.cs#L22). Generally, I use a listener if I need to do a computation that doesn't return an explicit result, but as a side-effect of the traverse, e.g., build a symbol table.

See https://tomassetti.me/listeners-and-visitors/

But, there are many ways to traverse the tree. In theory, you could visit any node in any order. XPath expressions define non-DFS traversals. For example, if one wants to find semantic predicates in an Antlr4 grammar, [we look for an "action" followed by a "?"](https://github.com/kaby76/g4-scripts/blob/c01e746656bf93b387063503ef9c01296bafa9e4/find-semantic-predicate.sh#L16C55-L16C98) in the parse tree.
